stubs/mu-plugin.php

```diff
@@ -0,0 +1,144 @@
+<?php
+/*
+ * Plugin Name: CacheWerk k6 (MU)
+ * Plugin URI: https://github.com/cachewerk/k6
+ * Description: ...
+ * Version: 1.0.0
+ * Author: Rhubarb Group
+ * Author URI: https://rhubarb.group
+ * Requires PHP: 7.2
+ */
+
+defined('ABSPATH') || exit;
+
+add_action('muplugins_loaded', ['k6ObjectCacheMetrics', 'init']);
+add_action('shutdown', ['k6ObjectCacheMetrics', 'print'], PHP_INT_MAX);
+
+class k6ObjectCacheMetrics
+{
+    const None = 'none';
+
+    const ObjectCachePro = 'object-cache-pro';
+
+    const RedisObjectCache = 'redis-cache';
+
+    const WpRedis = 'wp-redis';
+
+    const LiteSpeedCache = 'litespeed-cache';
+
+    protected static $cache;
+
+    protected static $client;
+
+    public static function init(): void
+    {
+        global $wp_object_cache;
+
+        if (! wp_using_ext_object_cache()) {
+            self::$cache = self::None;
+            self::$client = 'array';
+        }
+
+        $class = get_class($wp_object_cache);
+
+        if (strpos($class, 'RedisCachePro') === 0) {
+            self::$cache = self::ObjectCachePro;
+            self::$client = strtolower($wp_object_cache->clientName());
+        }
+
+        if (method_exists($class, 'connect_using_relay')) {
+            self::$cache = self::RedisObjectCache;
+
+            $client = class_exists('Redis') ? 'phpredis' : 'predis';
+            self::$client = defined('WP_REDIS_CLIENT') ? str_replace('pecl', 'phpredis', WP_REDIS_CLIENT) : $client;
+        }
+
+        if (method_exists($class, '_connect_redis')) {
+            self::$cache = self::WpRedis;
+            self::$client = 'phpredis';
+        }
+
+        if (property_exists($class, '_object_cache')) {
+            self::$cache = self::LiteSpeedCache;
+            self::$client = 'phpredis';
+        }
+    }
+
+    public static function print(): void
+    {
+        if (! self::$cache) {
+            return;
+        }
+
+        if (! self::shouldPrint()) {
+            return;
+        }
+
+        printf(
+            "\n<!-- plugin=%s client=%s %s -->\n",
+            self::$cache,
+            self::$client,
+            self::getMetrics()
+        );
+    }
+
+    protected static function shouldPrint(): bool
+    {
+        return ! ((defined('\WP_CLI') && constant('\WP_CLI')) ||
+                  (defined('\REST_REQUEST') && constant('\REST_REQUEST')) ||
+                  (defined('\XMLRPC_REQUEST') && constant('\XMLRPC_REQUEST')) ||
+                  (defined('\DOING_AJAX') && constant('\DOING_AJAX')) ||
+                  (defined('\DOING_CRON') && constant('\DOING_CRON')) ||
+                  (defined('\DOING_AUTOSAVE') && constant('\DOING_AUTOSAVE')) ||
+                  (function_exists('wp_is_json_request') && wp_is_json_request()) ||
+                  (function_exists('wp_is_jsonp_request') && wp_is_jsonp_request()));
+    }
+
+    protected static function getMetrics(): string
```

Comment:
Let's talk about sampling `redis-cli INFO` for every request.

```
sample#redis-hits=65
sample#redis-misses=10
sample#redis-hit-ratio=86.7
sample#redis-ops-per-sec=14
sample#redis-evicted-keys=0
sample#redis-used-memory=6243632
sample#redis-used-memory-rss=13533184
sample#redis-memory-fragmentation-ratio=2.2
sample#redis-connected-clients=2
sample#redis-tracking-clients=2
sample#redis-rejected-connections=0
sample#redis-keys=398
sample#relay-hits=12
sample#relay-misses=25
sample#relay-hit-ratio=32.43
sample#relay-ops-per-sec=1
sample#relay-keys=66
sample#relay-memory-used=1525120
sample#relay-memory-total=268435456
sample#relay-memory-ratio=0.57
```

Reply:
@tillkruss I've added Redis sampling for ROC and WP Redis. LiteSpeed doesn't give access to the underlying Redis instance, so I can't retrieve the data. I could use reflection, but not sure about any performance implications there. Thoughts?

Reply:
You can use Reflection, it's totally fine if not called excessively. 

There is also: https://ocramius.github.io/blog/accessing-private-php-class-members-without-reflection/